source/ada/lsp-ada_completions.adb

```diff
@@ -107,6 +138,308 @@ package body LSP.Ada_Completions is
+            end loop;
+
+            Append (Res, Trimed_S (Cur_Index .. Trimed_S'Last));
+         end;
+
+         return To_String (Res);
+      end Decode_String;
+
+      ---------------------------
+      -- Find_Next_Placeholder --
+      ---------------------------
+
+      function Find_Next_Placeholder
+        (S     : String;
+         Start : Integer)
+         return Match_Location
+      is
+         Placeholder_Regexp : constant Pattern_Matcher :=
+           Compile ("FooBar_[0-9]+");
+         Matched : Match_Array (0 .. 0);
+      begin
+         Match
+           (Self       => Placeholder_Regexp,
+            Data       => S,
+            Matches    => Matched,
+            Data_First => Start);
+         return Matched (0);
+      end Find_Next_Placeholder;
+
+      -----------------------
+      -- Post_Pretty_Print --
+      -----------------------
+
+      function Post_Pretty_Print (S : String) return String
+      is
+         Res : Unbounded_String;
+      begin
+         --  Remove this condition when V725-006 is fixed
+         if S (S'Last) = Ada.Characters.Latin_1.LF then
+            Append (Res, S (S'First .. S'Last - 1));
+         else
+            Append (Res, S);
+         end if;
+
+         --  Add back snippet terminator
+         Append (Res, "$0");
+         return To_String (Res);
+      end Post_Pretty_Print;
+
+   begin
+      if Me_Formatting.Active
+        and then Result.insertText.Is_Set
+        and then Result.insertTextFormat.Is_Set
+        and then Result.insertTextFormat.Value = Snippet
+      then
+         declare
+            Input : Utils.Char_Vectors.Char_Vector;
+            Output : Utils.Char_Vectors.Char_Vector;
+            PP_Messages : Pp.Scanner.Source_Message_Vector;
+            S : GNAT.Strings.String_Access;
+            Tmp_Unit : Analysis_Unit;
+            Cmd : Utils.Command_Lines.Command_Line := Context.Get_PP_Options;
+            Tmp_Context : constant Analysis_Context :=
+              Create_Context
+                (Charset =>
+                   Utils.Command_Lines.Common.Wide_Character_Encoding (Cmd));
+         begin
+            Set_PP_Switches (Cmd);
+
+            S :=
+              new String'
+                (VSS.Strings.Conversions.To_UTF_8_String
+                   (Result.insertText.Value));
+
+            declare
+               Full : constant String := Prefix & Encode_String (S.all);
+            begin
+               Input.Append (Full);
+               Tmp_Unit :=
+                 Get_From_Buffer
+                   (Context  => Tmp_Context,
+                    Filename => "",
+                    Buffer   => Full,
+                    Rule     => Rule);
+               --  -2 because:
+               --  - align using the column before the cursor
+               --  - and "column = offset - 1"
+               Pp.Actions.Set_Partial_Gnatpp_Offset
+                 (Integer (Span.first.character) - 2);
+               Pp.Actions.Format_Vector
```

Comment:
I think there is a new API for partial formatting, that simplifies things. See https://git.adacore-it.com/c/libadalang-tools/+/119537/2/src/laltools-partial_gnatpp.ads

Let's try to use it id possible